### PR TITLE
Convert sse calls in client from async to sync

### PR DESCRIPTION
**Pull request by abidlabs**
This converts all of the HTTP / SSE calls in the python client from async to sync. There was no benefit in keeping them async as we would convert them to synchronous calls using `utils.synchronize_async` anyways and the use of this method was preventing the client from being used in environments that have their own async event management like gunicorn/gevent.

Closes: https://github.com/gradio-app/gradio/issues/7581

To test, create a file called `test.py` with the following code:

```py
from gevent import monkey
monkey.patch_all()
from gradio_client import Client
from flask import Flask

app = Flask(__name__)

client = Client("abidlabs/en2fr")  # a Space that translates from English to French

@app.route("/gen")
def gen():
      result = client.predict(
                  "hello",	
                  api_name="/predict"
              )
      return result

if __name__ == "__main__":
      app.run(host="0.0.0.0", port=5000)

```

Then run:

`gunicorn -k gevent -w 1 test:app --bind 0.0.0.0:5000`

Then visit `http://localhost:5000/gen` and you should see a valid greeting. 

Note: I didn't change the ws calls since websocket.send/receive are async by nature and so we'd have to do a workaround like `utils.synchronize_async` anyways to support them. So if you use the Client to connect to Gradio apps running 3.x  ("compatibility mode"), it will still use the async client.

### CODE REVIEW

File: client/python/gradio_client/utils.py [hunk 1]
```diff
@@ -243,6 +244,7 @@ class Communicator:
     reset_url: str
     should_cancel: bool = False
     event_id: str | None = None
+    prediction_complete: bool = False
```
Comment (abidlabs) [inline client/python/gradio_client/utils.py:247]: This field is added so that we can detect when a prediction is complete in the `stream_sse_v1plus()` thread so that we can check this field from the `check_for_cancel()` thread and have it terminate itself (since threads can't just be canceled like asyncio tasks)

Comment (abidlabs) [reply client/python/gradio_client/utils.py]: Update: this field has been renamed to `thread_complete`

File: client/python/test/test_client.py [hunk 2]
```diff
@@ -1240,7 +1240,7 @@ def test_download_private_file(self, gradio_temp_dir):
             src="gradio/zip_files",
         )
         url_path = "https://gradio-tests-not-actually-private-spacev4-sse.hf.space/file=lion.jpg"
-        file = client.endpoints[0]._upload_file(url_path)  # type: ignore
+        file = client.endpoints[0]._upload_file(url_path, 0)  # type: ignore
```
Comment (abidlabs) [inline client/python/test/test_client.py:1243]: This test is not related to this but needed to be fixed since we changed the signature of `_upload_file`. It wasn't caught before because this test is marked flaky

**Review by freddyaboulton — APPROVED** [review]
This looks good to me @abidlabs! Tested some local demos and everything looks good. One comment I have is whether we can stop raising `CancelledError` and just return now? That should stop the job in the threadpool no?

File: client/python/gradio_client/utils.py [hunk 3]
```diff
@@ -311,7 +313,7 @@ async def get_pred_from_ws(
                             # otherwise will get nasty warning in console
                             task.cancel()
                             await asyncio.gather(task, reset, return_exceptions=True)
-                        raise CancelledError()
+                        raise concurrent.futures.CancelledError()
```
Comment (freddyaboulton) [inline client/python/gradio_client/utils.py:316]: Why this change?

File: client/python/gradio_client/utils.py [hunk 4]
```diff
@@ -400,58 +398,55 @@ async def get_pred_from_sse_v1plus(
...
-    )
-
-    for task in pending:
-        task.cancel()
-        try:
-            await task
-        except asyncio.CancelledError:
-            pass
+    helper.thread_complete = False
+    with concurrent.futures.ThreadPoolExecutor() as executor:
```
Comment (freddyaboulton) [inline client/python/gradio_client/utils.py:402]: Can we use the same one used in the parent `Client` to not have to create another one?

**Review by aliabid94 — APPROVED** [review]
tested, LGTM!

**Comment by abidlabs** [comment]
> This looks good to me @abidlabs! Tested some local demos and everything looks good. One comment I have is whether we can stop raising CancelledError and just return now? That should stop the job in the threadpool no?

Yup I think you're right

File: client/python/gradio_client/utils.py (hunk 3 above)
Comment (abidlabs) [reply client/python/gradio_client/utils.py:316]: This is just because I've replaced

```
from concurrent.futures import CancelledError
```

with 

```
import concurrent.futures
```

**Comment by abidlabs** [comment]
> One comment I have is whether we can stop raising CancelledError and just return now? That should stop the job in the threadpool no?

Ok so I spent some time on this, but if we stop raising CancelledError and instead return `None`, then several upstream methods fail. The basic problem is we need to decide what `job.result()` should produce if the `job` is canceled. I think its not a bad idea for it to just raise a `CancelledError` as is currently the behavior? We could potentially change this and I'm open to other ideas, but I think we can make that part of another PR, perhaps as we address: https://github.com/gradio-app/gradio/issues/8188

**Comment by abidlabs** [comment]
I'll go ahead and merge this in but open to suggestions on this @freddyaboulton

**Comment by abidlabs** [comment]
Thanks for the reviews @aliabid94 and @freddyaboulton!

**Comment by kxbin** [comment]
great work :+1:
